cmd/cleanup.go

```diff
@@ -0,0 +1,116 @@
+func (cmd *CleanupCmd) SetFlags(f *flag.FlagSet) {
+	f.StringVar(&cmd.jobId, "jobId", "", "Flag for specifying the migration jobId")
+	f.StringVar(&cmd.dataShardIds, "dataShardIds", "", "Flag for specifying a comma separated list of dataShardIds to be cleaned up. Defaults to ALL shards. Optional flag, and only valid for a sharded migration.")
+	f.StringVar(&cmd.targetProfile, "target-profile", "", "Flag for specifying project and instance details of Spanner e.g., \"project=XYZ,instance=ABC\"")
+	f.BoolVar(&cmd.datastream, "datastream", false, "Flag for specifying if Datastream streams associated with the migration job should be cleaned up or not. Defaults to FALSE.")
+	f.BoolVar(&cmd.dataflow, "dataflow", false, "Flag for specifying if Dataflow job associated with the migration job should be cleaned up or not. Defaults to FALSE.")
+	f.BoolVar(&cmd.pubsub, "pubsub", false, "Flag for specifying if pubsub associated with the migration job should be cleaned up or not. Defaults to FALSE.")
+	f.BoolVar(&cmd.monitoring, "monitoring", false, "Flag for specifying if monitoring dashboards associated with the migration job should be cleaned up or not. Defaults to FALSE.")
+	f.StringVar(&cmd.logLevel, "log-level", "DEBUG", "Configure the logging level for the command (INFO, DEBUG), defaults to DEBUG")
+	f.BoolVar(&cmd.validate, "validate", false, "Flag for validating if all the required input parameters are present")
+}
+
+func (cmd *CleanupCmd) Execute(ctx context.Context, f *flag.FlagSet, _ ...interface{}) subcommands.ExitStatus {
+	err := logger.InitializeLogger(cmd.logLevel)
+	if err != nil {
+		fmt.Println("Error initialising logger, did you specify a valid log-level? [DEBUG, INFO, WARN, ERROR, FATAL]", err)
+		return subcommands.ExitFailure
+	}
+	targetProfile, err := profiles.NewTargetProfile(cmd.targetProfile)
+	if err != nil {
+		logger.Log.Debug(fmt.Sprintf("Target profile is not properly configured, this is needed for SMT to lookup job details in the metadata database: %v\n", err))
+		return subcommands.ExitFailure
+	}
+	project, instance, err := streaming.GetInstanceDetails(ctx, targetProfile)
+	if err != nil {
+		logger.Log.Debug(fmt.Sprintf("can't get resource ids: %v\n", err))
+		return subcommands.ExitFailure
+	}
+	dataShardIds, err := profiles.ParseList(cmd.dataShardIds)
+	if err != nil {
+		logger.Log.Debug(fmt.Sprintf("Could not parse datashardIds: %v\n", err))
+		return subcommands.ExitFailure
+	}
+	if !(cmd.datastream || cmd.dataflow || cmd.pubsub || cmd.monitoring) {
+		logger.Log.Error("At least one of datastream, dataflow, pubsub or monitoring must be specified, we recommend cleaning up all resources!\n")
+		return subcommands.ExitUsageError
+	}
+	// all input parameters have been validated
+	if cmd.validate {
+		logger.Log.Info("All input parameters are valid.")
+		return subcommands.ExitSuccess
+	}
+	jobExecutionData, shardExecutionDataList, err := streaming.GetJobDetails(ctx, cmd.jobId, dataShardIds, project, instance)
```

Comment:
We don't seem to be storing anywhere the details of migration type i.e. POC or minimal . And the job details are only populated for streaming migrations, in future when we aim to add statefulness to these migrations we might need this detail captured somewhere

Reply:
No statefulness for POC migrations

Reply:
sure, makes sense